src/datumaro/util/mask_tools.py

```diff
@@ -124,8 +124,54 @@ def remap_mask(mask, map_fn):
     return np.array([map_fn(c) for c in range(256)], dtype=np.uint8)[mask]
 
 
-def make_index_mask(binary_mask, index, dtype=None):
-    return binary_mask * np.array([index], dtype=dtype or np.min_scalar_type(index))
+def make_index_mask(
+    binary_mask: np.ndarray,
+    index: int,
+    ignore_index: int = 0,
+    dtype: Optional[np.dtype] = None,
+):
+    """Create an index mask from a binary mask by filling a given index value.
+
+    Args:
+        binary_mask: Binary mask to create an index mask.
+        index: Scalar value to fill the ones in the binary mask.
+        ignore_index: Scalar value to fill in the zeros in the binary mask.
+            Defaults to 0.
+        dtype: Data type for the resulting mask. If not specified,
+            it will be inferred from the provided index. Defaults to None.
+
+    Returns:
+        np.ndarray: Index mask created from the binary mask.
+
+    Raises:
+        ValueError: If dtype is not specified and incompatible scalar types are used for index
+            and ignore_index.
+
+    Examples:
+        >>> binary_mask = np.eye(2, dtype=np.bool_)
+        >>> index_mask = make_index_mask(binary_mask, index=10, ignore_index=255, dtype=np.uint8)
+        >>> print(index_mask)
+        array([[ 10, 255],
+               [255,  10]], dtype=uint8)
+    """
+    if dtype is None:
+        dtype = np.min_scalar_type(index)
+        if dtype != np.min_scalar_type(ignore_index):
+            raise ValueError()
+
+    flipped_zero_np_scalar = ~np.full(tuple(), fill_value=0, dtype=dtype)
+
+    # NOTE: This dispatching rule is required for a performance boost
+    if ignore_index == flipped_zero_np_scalar:
+        flipped_index = ~np.full(tuple(), fill_value=index, dtype=dtype)
+        return ~(binary_mask * flipped_index)
```

Comment:
Sorry, I was failed to understand this logic correctly. When `ignore_index` is 255, this returns `binary_mask * np.full(tuple(), fill_value=index, dtype=dtype)`. Is this right?
There is two negative `~` operations. Should we have this?
We have to explanation what `ignore_index==255` stands for.

Reply:
Please see this example
https://github.com/openvinotoolkit/datumaro/blob/53b3a5aa02d3e1062a00e68e2e8058de18c0a62d/src/datumaro/util/mask_tools.py#L152-L157

We want to convert a binary mask to an index mask by filling `1`s with `index` and `0`s with `ignore_index`. This function would be a simple function like
```python
return np.where(binary_mask, mask, ignore_index)
```
, but this function has if-else branching in the loop. Thus, it is hard to be accelerated by hardware.

`~` is a bit-wise flipping operator. Therefore, this logic is doing (`~255 == 0`)
```python
[~10, 0]
[0, ~10]

=>

[10, 255]
[255, 10]
```

Anyway, https://github.com/openvinotoolkit/datumaro/pull/1409/commits/53b3a5aa02d3e1062a00e68e2e8058de18c0a62d improved it further to cover the general cases.

Reply:
```python
def kernel4(binary_mask, index, ignore_index=0, dtype=None):
    if dtype is None:
        dtype = np.min_scalar_type(index)
        if dtype != np.min_scalar_type(ignore_index):
            raise ValueError()

    mask = binary_mask * np.array([index], dtype=dtype)

    if ignore_index == 0:
        return mask

    return np.where(binary_mask, mask, ignore_index)

from timeit import timeit

binary_mask = np.random.randint(0, 2, size=[1024, 1024], dtype=np.bool_)

print("Kernel4 time: ", timeit(lambda: kernel4(binary_mask, index=10, ignore_index=230), number=200))
print("make_index_mask time: ", timeit(lambda: make_index_mask(binary_mask, index=10, ignore_index=230), number=200))
```
There was a huge performance gap with/without this optimization in my desktop
```console
Kernel4 time:  0.8781529648695141
make_index_mask time:  0.03690833714790642
```